titus-common/src/main/java/com/netflix/titus/common/util/archaius2/SpringProxyInvocationHandler.java

```diff
@@ -0,0 +1,216 @@
+/*
+ * Copyright 2020 Netflix, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.titus.common.util.archaius2;
+
+import java.lang.reflect.InvocationHandler;
+import java.lang.reflect.Method;
+import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.Proxy;
+import java.lang.reflect.Type;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+
+import com.google.common.base.Preconditions;
+import com.netflix.archaius.api.annotations.Configuration;
+import com.netflix.archaius.api.annotations.DefaultValue;
+import com.netflix.archaius.api.annotations.PropertyName;
+import com.netflix.titus.common.util.StringExt;
+import com.netflix.titus.common.util.tuple.Either;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.core.env.Environment;
+
+class SpringProxyInvocationHandler implements InvocationHandler {
+
+    private static final Logger logger = LoggerFactory.getLogger(SpringProxyInvocationHandler.class);
+
+    private interface MethodHandler {
+        Object get() throws Throwable;
+    }
+
+    private final Class<?> apiInterface;
+    private final String prefix;
+    private final Environment environment;
+
+    private final Map<Method, MethodHandler> methodWrappers;
+
+    SpringProxyInvocationHandler(Class<?> apiInterface, String prefix, Environment environment) {
+        Preconditions.checkArgument(apiInterface.isInterface(), "Not interface: %s", apiInterface);
+
+        this.apiInterface = apiInterface;
+
+        String effectivePrefix = prefix;
+        if (prefix == null) {
+            Configuration configurationAnnotation = apiInterface.getAnnotation(Configuration.class);
+            effectivePrefix = configurationAnnotation != null ? configurationAnnotation.prefix() : null;
+        }
+        this.prefix = StringExt.isEmpty(effectivePrefix) ? "" : (effectivePrefix.endsWith(".") ? effectivePrefix : effectivePrefix + '.');
+        this.environment = environment;
+
+        Map<Method, MethodHandler> methodWrappers = new HashMap<>();
+        for (Method method : apiInterface.getMethods()) {
+            Preconditions.checkArgument(method.getParameterCount() == 0, "Method with no parameters expected");
```

Comment:
wouldn't it be safer to just ignore this method and consider it not a configurable property? Especially as we are looking at all methods, including the inherited ones from parent interfaces.

Reply:
We cannot just ignore them. We could either:
1. execute them if a default implementation is provided (our type is interface not a class)
2. throw an exception if the method is called by the client, but do not fail eagerly in the constructor